feat(InvoiceNinja Node): extended support for most of the available resources, operations and events in v5

paulwer opened the pull request:
Changes to the node implementing current invoiceninja methods and events. Some things, like methods for CRUD projects or vendors are still missing.

Event-IDs are read by invoiceninja database administration.

Documentation should be after merging.

paulwer committed:
updated node invoice-ninja with new events and functions

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/n8n-io/n8n?pullRequest=4807) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/n8n-io/n8n?pullRequest=4807) before we can accept your contribution.<br/><hr/>**Paul Werner** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/n8n-io/n8n?pullRequest=4807) it.</sub>

paulwer committed:
adding new delete webhook

paulwer committed:
fixing wrong event names for quotes

paulwer commented:
There are no breaking changes involved. Do we need an updated version-number of the node here?

paulwer commented:
@RicardoE105 Do you need any further informations on this?

paulwer committed:
fixing wrong event names

paulwer committed:
fixing wrong event name

paulwer committed:
adding webhook update_project

paulwer committed:
adding new variables to interfaces and to handlers + changes order of methods

paulwer commented:
there are some mayor changes between v4 and v5. I've tried to update the node ready for v5 endpoints, but encountered many fields, which does not exists on v4 and are present within the old code.

Did someone know, if they come from v4 and therefore should not be deletet to downward-compatibility?

paulwer committed:
Merge https://github.com/n8n-io/n8n into feature-invoice-ninja

paulwer commented:
new webhooks are tested on a live version
enpoints are tested each and partialy for the fields

paulwer committed:
adding newly added update_payment webhook

paulwer commented:
i will split the codebase for v4 and v5 more strictly to ensure backward compatibility.

paulwer commented:
@Joffcom
Is there a way, to split the Descriptions based on a node-parameter, like here called "apiVersion"?

If not, I would like to split the current node into 2 nodes. (Invoice Ninja V4 & Invoice Ninja V5), because the differ to much from one another and are hardly to manage both at the same time.
I've tried to split them based on the Description parameters, but it seems not to integrate well into the n8n system. (like attached)  

Please present me the optimal next steps, to make this separation happen. Is there a way to deprecate the existing node and create 2 separate new Nodes? Or do you prefer anything else?
[InvoiceNinja.zip](https://github.com/n8n-io/n8n/files/10499594/InvoiceNinja.zip)

Joffcom commented:
Hey @paulwer,

There shouldn't be any need to make 2 nodes as we support v4 and v5 in the node through the node versioning, If you have a setting that only exists for v5 you can set the display the display to only show for that version and the same for any data structures that are different you can use the `apiVersion` to work out what endpoint to use.

If you look at the current node you can see where we are already doing this.

paulwer committed:
split codebase + other features
- split by versions
- using interfaces from main repo for v5
- new ressources (project, vendor)
- jsonBody Parameter for create/update resources
- parameter checks done for: projects, clients, expenses
(others are pending)

paulwer committed:
WIP: readding parameters & functionality for fetching data
(getAll)

paulwer committed:
Merge remote-tracking branch 'upstream/master' into feature-invoice-ninja

paulwer committed:
reworked pagination

paulwer committed:
fixes

paulwer committed:
download invoices + updates to pagination

paulwer committed:
adding Product Ressource

paulwer committed:
update ressource: invoice

paulwer committed:
introduce versioned node, for backwards compatibility

paulwer committed:
note for migrating to IN V5

paulwer committed:
adding new webhooks

paulwer committed:
adding vendor webhooks

paulwer committed:
adding resource: recurring invoices

paulwer committed:
new ressource: recurring expenses + fixes

paulwer commented:
@Joffcom i've implemented the node versioning pattern from HTTPRequest

Please check and may give some feedback

paulwer committed:
cycle params for recurring invoices / expenses

paulwer committed:
update ressource: quote

paulwer committed:
updated ressource: payment + fixes for clientId & projectId fields

paulwer committed:
update ressource: vendors

paulwer committed:
new ressource: credit

paulwer committed:
Merge remote-tracking branch 'upstream/master' into feature-invoice-ninja

Joffcom commented:
Hey @paulwer,

That is a massive amount of work there but maybe a little but over the top, We already had the light versioning enabled in the node which was used to allow support for v4 and v5 which I would have expected to work. It also allows users to change the version they are using from the n8n UI.

![image](https://user-images.githubusercontent.com/4688521/216111505-77c875d6-fd2b-4743-b8f2-d3cbf4a84589.png)

paulwer commented:
Hey @Joffcom,
the current node does only supports some of the api-features.
I'am currently developing this only for InvoiceNinja V5. The user still has the option to choose between v4 and v5 in the node. The code is just more split, than before, because some of the fields are called differently or neighter exists. 

Because, there are some mayor breaking changes (other paramater names in n8n), I had implemented the versioning of the whole node, like in the HTTPRequest node. (dont get confused, between versions of N8N and versions of InvoiceNinja)

There are 4 more ressources to come.

paulwer commented:
@Joffcom 

When a user enables apiVersion v4 they can do the same operations, like before, for the givven resources.
When they switch/migrate to v5, they can use even more resources/operations, with focus on the v5 of InvoiceNinja.

If you have the time, checkout my fork and run the node to view the actual aproach.

Resources within v4: (as already implemented)

Resources within v5: (partialy new and not available for v4-users)

Joffcom commented:
Hey @paulwer,

The HTTP Request node versioning is more for what we call Full Versioning which would apply for a complete node overhaul which would be an internal task for when we completely change the functionality.

Different field names or new ones should have been achievable using what was there already but once you are finished we can create a ticket for getting this reviewed.

paulwer committed:
new ressource: bank transactions + updated download operations (quote, recuring invoices)

paulwer committed:
Merge remote-tracking branch 'upstream/master' into feature-invoice-ninja

paulwer commented:
@Joffcom imagine, if we are using the node versioning strategy, how does users are promted to update their nodes to the newest version?

paulwer commented:
@parasdaryanani 
info: I had implemented your change for this update within v5 [670efb9](https://github.com/n8n-io/n8n/pull/4807/commits/670efb9128f4557b09ec5f39b6f4210b9ac051d0)

paulwer committed:
missing changes of merge

paulwer committed:
adding download operation for credit resource

paulwer committed:
webhooks sent events + purchase_order events

paulwer committed:
add ressource PurchaseOrder + fixes + product events

paulwer committed:
fix: ts-error function to big for sentiment analysis + order imported objects within node

paulwer committed:
new ressource: subscription + fix at vendor

paulwer commented:
@Joffcom the main work is done, I will now just get an review by the invoiceninja team for the fields and will come back to your, after we had finished the testing and QS.

Please tell me in the meantime, how we want to process after this phase.

Joffcom commented:
Hey @paulwer,

I spotted the thread on the Invoice Ninja issues, I think on our side we would need to review the code and the structure to check the quality and that it can be easily maintained, passes the lint rules, doesn't have a lot of code-reuse and that all the node versions work.

As it adds more features we will need to look at updates to the docs and also see where Invoice Ninja is on our node overhaul plan as we are changing the design and standardising some of the inputs / fields that nodes have.

Because there is such a drastic change to this node though that may not have been needed rather than an hour or 2 to review this will likely take a lot longer so will need to go through a prioritisation process to see where it can fit between all of the other things we are currently doing.

I will of course let you know when we start the review.

paulwer committed:
fixes

paulwer commented:
Hey @Joffcom,
thanks for your reply. docs are already prepared: https://github.com/n8n-io/n8n-docs/pull/1033
Please reach out to me with more informations, when you have them. Do you have a time-frame, in which I can exspect the next action?

paulwer committed:
linting node

paulwer committed:
fixes (userId) not changeable

paulwer committed:
fixes: client

paulwer committed:
adding bulk actions for clients and vendors

paulwer committed:
more bulk actions + fixes + move download operation to get operation as an option + send email operations to invoice, quote, credit

Joffcom commented:
Hey @paulwer,

No timeframe at the moment, I have a few PRs in my queue ahead of this one but we will need to prioritise this one.

paulwer committed:
exclude payment type as Entity like Country Codes, removing includes from clients, changes to expense params, only get last 100 results for fetching some props for select

paulwer commented:
Hey @Joffcom,
thanks, so we are talkiing about weeks and not multiple months?
Would sounds great for me. I will take a note here, if the pr is reviewable.

Thank you kindly

Joffcom commented:
Hey @paulwer,

It all comes down to that priority and if anything urgent comes up before then so it could be weeks or it could slip to months. Ideally I would love to aim for weeks but I can't promise anything as things could change quickly.

Generally as a rule I spent time on Thursdays doing PR reviews but I may split that over a couple of days to get the count down.

paulwer committed:
fixes for invoice

paulwer committed:
api secret for downloading files

paulwer committed:
updated downloadFile function to use builtin Authentification + new ressource documents + fix of incorrect usage of "that" instead of "this" with fn.call(this)

paulwer commented on packages/nodes-base/nodes/InvoiceNinja/V3/v5/DocumentExecution.ts:
```diff
@@ -0,0 +1,111 @@
+            if (operation === 'upload') {
+                const uploadRessource = this.getNodeParameter('uploadRessource', i);
+                const uploadRessourceId = this.getNodeParameter('uploadRessourceId', i);
+                const binaryPropertyName = this.getNodeParameter('binaryPropertyName', i);
+                const binaryData: IBinaryData = items[i].binary![binaryPropertyName];
+                if (!binaryData) throw new NodeOperationError(this.getNode(), 'No binary data exists on item!', {
+                    itemIndex: i,
+                });
+                const formData = new FormData();
+                formData.append('documents[]', new Blob([binaryData.data], { type: binaryData.mimeType }), binaryData.fileName);
+                responseData = await invoiceNinjaApiRequestUploadFile.call(
+                    this,
+                    'POST',
+                    `/${uploadRessource}/${uploadRessourceId}/upload?_method=put`,
+                    formData,
+                );
+                responseData = responseData.data;
```
@Joffcom can you provide me details / examples, how I can send FormData with a file to the api.
The postman request looks like this:
![image](https://user-images.githubusercontent.com/52678724/217035287-e548fdb8-a184-49db-8926-fcabb9ac6642.png)

I've tried the function this.helpers.requestWithAuthentication and provided eigher property body or property formData. when providing formData to body it failes with the following error: Data after transformation must be a string, an ArrayBuffer, a Buffer, or a Stream. When providing formData as an own property, there was no affect at the ressource in the api, like I'have send no data.

Joffcom replied:
Hey @paulwer,

Have you tried checking the HTTP Request node which I think might do this?

paulwer replied:
@Joffcom I found a solution copied by mindee node.
I guess there was something wrong, how the binary was handled while form generation or how the form gets attached, because this solution feeld like an workaround.

Please considder this at your review and maybe improve this section.
If you need any assistence, please reach out.

paulwer committed:
fix: upload Documents + custom FileName while uploading document

paulwer committed:
fixes for payments, documents, clients + password routes (for protected ressources) with new credentials property

paulwer committed:
fixes for projects

paulwer committed:
fixes for quotes and projects + fixes for dates (wrong formats get ignored) + system endpoints

paulwer committed:
fixes

paulwer committed:
adding currency entity + checks for vendor

paulwer committed:
checks for products

paulwer committed:
adding client statement

paulwer committed:
mark sent

paulwer committed:
fixes for nullable parameters or parameters with negative boolean value

paulwer committed:
fixes

paulwer committed:
fix of new webhook behavior

paulwer committed:
checks for bank_transactions + other fixes

paulwer committed:
checks for credits + fixes for invoices,quotes and recurring invoices + replace btoa to Buffer syntax

paulwer commented:
@Joffcom I've runned the command lint:fix and the resources were ordered by alphanumeric.

is it wanted / possible to order the ressources like the order within the docs for use and topics?

paulwer committed:
checks for purchase_orders

paulwer committed:
fixes

paulwer committed:
fixes

paulwer committed:
check recurring expenses + frequency to entity

paulwer committed:
Merge remote-tracking branch 'upstream/master' into feature-invoice-ninja

o-psi commented:
@Joffcom A reworking of this node is needed anyway as the node does not work currently, correct?

We use N8N primarily with Invoice ninja, so its fairly important for us to have this working. 

@paulwer I am going to install your fork and begin testing.

paulwer commented:
@o-psi there was a bug within the latest version, of creating new webhooks, that should be fixed now.
https://github.com/invoiceninja/invoiceninja/issues/8260

paulwer committed:
checks for subscriptions + linting

paulwer commented:
@Joffcom @o-psi i am done with my tests.

the "problem" is, that we dont have any DTOs defined by InvoiceNinja, so I try to test every parameter if it have an effect for the result and removed these in the last updates while testing. I hoped, something has not slipped through.

paulwer committed:
fixed spelling of event-names

paulwer commented:
@o-psi any updates/feedback/wishes? have you did some testing activities?

paulwer committed:
Merge remote-tracking branch 'upstream/master' into feature-invoice-ninja

paulwer committed:
additional change change: feat: Migrate packages from lodash to lodash-es (no-changelog) (#5538)

paulwer commented:
@Joffcom @alexgrozav I included the latest change from https://github.com/n8n-io/n8n/pull/5538 into the pr.

paulwer committed:
Merge remote-tracking branch 'upstream/master' into feature-invoice-ninja

paulwer committed:
related changes to last merge

paulwer commented:
@Joffcom i updated the pr to the latest changes of the master branch.

please check the name-convention of the trigger, because i had to set an eslint disable here to ignore the accuring error.

paulwer commented on packages/nodes-base/nodes/InvoiceNinja/V1u2/InvoiceNinjaTriggerV1u2.node.ts:
```diff
@@ -0,0 +1,242 @@
+			defaults: {
+				name: 'Invoice Ninja Trigger',
+			},
+			inputs: ['main'],
```
@Joffcom eslint also mentioned a fix for this line: 
inputs: [] to inputs. ['main']

please check

paulwer commented on packages/nodes-base/nodes/InvoiceNinja/V3/InvoiceNinjaTriggerV3.node.ts:
```diff
@@ -0,0 +1,119 @@
+			defaults: {
+				name: 'Invoice Ninja Trigger',
+			},
+			inputs: ['main'],
```
same here :)

paulwer committed:
Merge remote-tracking branch 'upstream/master' into feature-invoice-ninja

paulwer committed:
additional changes to latest merge

Joffcom commented:
Hey @paulwer 

The review is not in progress yet as I still have a fairly large list of other requests to look at which can be seen by looking at the PRs with a community label.

Once it starts I will let you know but looking at how much this has changed there is a lot of testing to do internally and it could be that we drop the versioning in favour of the previously mentioned method with the light versioning so that users are able to swap between versions which is important if someone is likely to be self hosting.

paulwer committed:
Merge remote-tracking branch 'upstream/master' into feature-invoice-ninja

paulwer committed:
additional changes to merge

paulwer committed:
Merge remote-tracking branch 'upstream/master' into feature-invoice-ninja

paulwer committed:
Merge branch 'master' into feature-invoice-ninja

paulwer committed:
Merge branch 'master' into feature-invoice-ninja

paulwer committed:
Merge remote-tracking branch 'upstream/master' into feature-invoice-ninja

paulwer committed:
additional changes refering to #6015

Joffcom commented:
@paulwer not yet, once the review starts a note will be placed on here.

It looks like there are currently 31 community node improvements ahead of this one in the queue.

paulwer committed:
Merge remote-tracking branch 'upstream/master' into feature-invoice-ninja

paulwer commented:
sorry to bother you again, but i catched, that within the last 5 weeks nothing has changed at the remaining count of PR's ahead of this one.
Is there any chance to prioritise this one, like suggesed within a message of you in february?

would like to wrap it up^^
if there are any task, which could be done just reach out :)

Joffcom commented:
Hey @paulwer 

In the last 5 weeks there has been a lot of other things that have been checked and fixed, I am not sure if you have noticed but the amount of open issues has gone from 150+ to under 100.

Something that is broken will often have a higher priority than a new feature set.

I will have a quick initial look over this during my next PR review slot to see if it generally makes sense or if we will need to spend some time undoing parts of it.

paulwer committed:
Merge remote-tracking branch 'upstream/master' into feature-invoice-ninja

paulwer committed:
#6450 related

Joffcom commented:
Alright I have taken a quick look, It looks like the Trigger node and the app node are no longer being treated as the same node in the UI so it appears twice, The trigger node was also versioned although I suspect that isn't actually needed.

Looking at the node we currently have we toggle between v4 and v5 under the settings which can also impact UI fields but now I can see the node itself I can see the extra fields and I get that it was just easier to create a new version but it is not needed and the v2 of the node also contains the same changes so it seems a bit over the top.

The `Additional JSON Body` that appears on all of the operations can probably be hidden behind an additional fields option (which is normally the last option on a node).

The notice is nice but maybe not needed and the V4 / V5 next to the resource can also be removed, Looking at the changes that need to be made to it this it will probably take a day or 2.

The actual implementation of the API options and fields looks to be mostly ok it is just the way the versioning has been handled that needs to be sorted so that fits our current patterns. The downside here though is this node currently sits outside of our top 100 so getting any prioritisation on making these changes will be tricky and possibly unlikely so I may have to work on this between other issues to bring it back inline.

paulwer commented:
if there are any guidelines for the versioning you can provide I can take the initial step in reworking the code.
feeld free to reach out to me if you can give me a specific task on this pr.

:)

paulwer commented:
@Joffcom pls give some informations about the next steps asked above :)

Joffcom commented:
Hey @paulwer,

I will need to have a think and look back through previous comments I think we talked about this briefly a few months back, I would probably take the current node from the master branch and just keep the same light node versioning and add in the newer features.

paulwer commented:
if there is anything to do please reach out :)

paulwer commented:
@Joffcom sadly another month has nearly passed again. 
Please provide a plan, how we can wrap this up quickly. As I mentioned, I will assist.

I will probably be unavailable from 08.09.23 to nearly end of the year. It would be great if we can end this before this time!

Joffcom commented:
Hey @paulwer,

I currently do not have a timeline for this PR, I have now set up 2 slots each week for reviewing PRs but the order depends on when the PR came in and the complexity required to review.

As this PR requires a lot more than a quick review it will take longer, the problem is trying to fit something of this size into a normal review slot when we typically reserve this kind of thing for a node overhaul.

My plan for the node is to take the current node from the master branch then add in the changes to the light version with out all the extra complexity.

paulwer commented:
sounds like a plan. 👍
as mentioned, just reach out if you need any assistance.
i can also provide you with a test-instance of invoiceninja if you like to run some tests.

Joffcom committed:
Merge branch 'master' of github.com:n8n-io/n8n into feature-invoice-ninja

paulwer commented:
any updates?

Joffcom commented:
Hey @paulwer,

I am still working my way through it between other commitments and issues.

I have some of the changes in place but due to the scale it is still tricky to fit this in, Sadly this PR got way out of hand and didn't follow earlier recommendations which is part of why it is taking so long to complete.

paulwer commented:
if it helps, i can revisit with the IN devs to maybe drop v4-support and only support v5 of IN, if there is not so many people using it.

Ok for you? does this help?

Joffcom commented:
Hey @paulwer,

That doesn't really make a difference as it can be self hosted so just removing it would be a breaking change.

Even with that there is still some work needed to bring it up, it will get there though.

paulwer commented:
if you have any idea how I can help please mention it! I also want to conclude this. We are near the 1 year mark🙈

paulwer commented:
@Joffcom the docs-pr was closed

Joffcom commented:
Hey @paulwer,

You are not wrong there this has been opened for a lot longer than I expected from the initial version all those months ago.

Sadly though the scope of the changes made and the direction taken just hasn't made this easy or quick to fit in around everything else.

Normally something of this size would be closed with a note to break it up into smaller chunks and to follow the pattern of the current node.

We might have been ok with the first few commits before it went off the rails then we could have had a chat about to overhaul the node properly.

The issue right now is I just lack the time to fix and adjust everything so I am having to fit it in between other tasks.

If you wanted to take a bash at it we need to revert the entire second version you created and fit it into the current light version then we can think about if it is worth doing a full version using newer ui components.

paulwer commented:
Can you explain the next steps a bit further.

Joffcom commented:
@paulwer of course...

So as a step 1 we revert all changes then update the current light version to include some of the missing options this allows users still self hosting v4 to use and select that as a version with no issues.

Then we can look at making a full version just for v5 that uses the same code style and patterns we use in other nodes as at the moment this node doesn't follow our pattern at all so needs to be fully reworked for this.

paulwer commented:
@Joffcom @CodeShakingSheep i am sadly not able to work on this in my current situation. feel free to reuse what you can :)

ShireenMissi commented:
Thank you so much for your contribution! 🙏 
This PR has been inactive for a while, so I’m going to close it for now. If you’d like to revisit it in the future, please feel free to open a new PR based on the current codebase. We’d be happy to review it.
